worldedit-bukkit/src/main/java/com/boydti/fawe/bukkit/adapter/mc1_16_5/TuinityRelighter_1_16_5.java

```diff
@@ -0,0 +1,130 @@
+package com.boydti.fawe.bukkit.adapter.mc1_16_5;
+
+import com.boydti.fawe.Fawe;
+import com.boydti.fawe.beta.IQueueChunk;
+import com.boydti.fawe.beta.IQueueExtent;
+import com.boydti.fawe.beta.implementation.chunk.ChunkHolder;
+import com.boydti.fawe.beta.implementation.lighting.Relighter;
+import com.boydti.fawe.util.TaskManager;
+import net.minecraft.server.v1_16_R3.BlockPosition;
+import net.minecraft.server.v1_16_R3.ChunkCoordIntPair;
+import net.minecraft.server.v1_16_R3.LightEngineThreaded;
+import net.minecraft.server.v1_16_R3.MCUtil;
+import net.minecraft.server.v1_16_R3.WorldServer;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import java.util.function.IntConsumer;
+
+public class TuinityRelighter_1_16_5 implements Relighter {
+
+    private static final IntConsumer nothingIntConsumer = i -> {};
+    private static final MethodHandle relight;
+
+    private final WorldServer world;
+    private final ReentrantLock lock = new ReentrantLock();
+
+    private final IQueueExtent<IQueueChunk> queue;
+
+    static {
+        MethodHandle tmp = null;
+        try {
+            Method relightMethod = LightEngineThreaded.class.getMethod(
+                    "relight",
+                    Set.class,
+                    Consumer.class,
+                    IntConsumer.class
+            );
+            tmp = MethodHandles.lookup().unreflect(relightMethod);
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+            e.printStackTrace();
+        }
+        relight = tmp;
+    }
+
+    public TuinityRelighter_1_16_5(WorldServer world, IQueueExtent<IQueueChunk> queue) {
+        this.queue = queue;
+        this.world = world;
+    }
+
+    @Override
+    public boolean addChunk(int cx, int cz, byte[] skipReason, int bitmask) {
+        List<ChunkCoordIntPair> chunks = MCUtil.getSpiralOutChunks(new BlockPosition(cx << 4, 0, cz << 4), 1);
+        TaskManager.IMP.task(() ->
+                {
+                    try {
+                        // cast is required for invokeExact
+                        int unused = (int) relight.invokeExact(world.getChunkProvider().getLightEngine(),
+                                (Set<?>) new HashSet<>(chunks), // explicit cast to make invokeExact work
+                                (Consumer<?>) coord -> sendChunk(cx, cz), // send chunk after lighting was done
```

Comment:
the light engine will resend lighting on its own, do not send chunks here. especially in an async context

Reply:
When not sending chunks here, changes aren't visible to the player. Not sure why, I basically copied over the behaviour of the NMSRelighter here.

Reply:
It's ok to send the chunks since ligthing part is ignored by the third param ``false``. In this case, ``PacketPlayOutMapChunk`` is the only one to be sent to players.

Reply:
Please test this thoroughly to make sure this works without any issues.

Reply:
Also, mark it as resolved once you are sure @Spottedleaf's comment has been addressed.

Reply:
The third parameter for `PacketPlayOutMapChunk` is actually controlling paper's anti-xray behavior, nothing about lighting. Lighting is always sent as the light update packet in 1.14+.

Reply:
Since light updates and block updates are separate, if you are using this to actually also re-send block changes then I guess it's fine. I would suggest posting the send to the main thread, for obvious reasons.